src/vs/editor/browser/gpu/fullFileRenderStrategy.ts

```diff
@@ -0,0 +1,371 @@
+
+
+				for (x = tokenStartIndex; x < tokenEndIndex; x++) {
+					// HACK: Prevent rendering past the end of the render buffer
+					// TODO: This needs to move to a dynamic long line rendering strategy
+					if (x > FullFileRenderStrategy._columnCount) {
+						break;
+					}
+					chars = content.charAt(x);
+					if (chars === ' ') {
+						continue;
+					}
+					if (chars === '\t') {
+						// TODO: Pull actual tab size
+						xOffset += 3;
+						continue;
+					}
+
+					glyph = this._atlas.getGlyph(this._glyphRasterizer, chars, tokenMetadata);
+
+					// TODO: Support non-standard character widths
+					screenAbsoluteX = Math.round((x + xOffset) * viewLineOptions.spaceWidth * activeWindow.devicePixelRatio);
+					screenAbsoluteY = (
+						Math.ceil((
+							// Top of line including line height
+							viewportData.relativeVerticalOffset[y - viewportData.startLineNumber] +
+							// Delta to top of line after line height
+							Math.floor((viewportData.lineHeight - this._context.configuration.options.get(EditorOption.fontSize)) / 2)
+						) * activeWindow.devicePixelRatio)
+					);
+					zeroToOneX = screenAbsoluteX / this._canvas.width;
+					zeroToOneY = screenAbsoluteY / this._canvas.height;
+					wgslX = zeroToOneX * 2 - 1;
+					wgslY = zeroToOneY * 2 - 1;
+
+					cellIndex = ((y - 1) * FullFileRenderStrategy._columnCount + (x + xOffset)) * Constants.IndicesPerCell;
+					cellBuffer[cellIndex + CellBufferInfo.Offset_X] = wgslX;
+					cellBuffer[cellIndex + CellBufferInfo.Offset_Y] = -wgslY;
+					cellBuffer[cellIndex + CellBufferInfo.GlyphIndex] = glyph.glyphIndex;
+					cellBuffer[cellIndex + CellBufferInfo.TextureIndex] = glyph.pageIndex;
+				}
+
+				tokenStartIndex = tokenEndIndex;
+			}
+
+			// Clear to end of line
+			fillStartIndex = ((y - 1) * FullFileRenderStrategy._columnCount + (tokenEndIndex + xOffset)) * Constants.IndicesPerCell;
+			fillEndIndex = (y * FullFileRenderStrategy._columnCount) * Constants.IndicesPerCell;
+			cellBuffer.fill(0, fillStartIndex, fillEndIndex);
+
+			upToDateLines.add(y);
+		}
+
+		const visibleObjectCount = (viewportData.endLineNumber - viewportData.startLineNumber + 1) * lineIndexCount;
+
+		// Only write when there is changed data
+		if (dirtyLineStart <= dirtyLineEnd) {
+			// Write buffer and swap it out to unblock writes
+			this._device.queue.writeBuffer(
+				this._cellBindBuffer,
+				(dirtyLineStart - 1) * lineIndexCount * Float32Array.BYTES_PER_ELEMENT,
+				cellBuffer.buffer,
+				(dirtyLineStart - 1) * lineIndexCount * Float32Array.BYTES_PER_ELEMENT,
+				(dirtyLineEnd - dirtyLineStart + 1) * lineIndexCount * Float32Array.BYTES_PER_ELEMENT
+			);
+		}
+
+		this._activeDoubleBufferIndex = this._activeDoubleBufferIndex ? 0 : 1;
+
+		this._visibleObjectCount = visibleObjectCount;
+		return visibleObjectCount;
+	}
+
+	draw(pass: GPURenderPassEncoder, viewportData: ViewportData): void {
+		if (this._visibleObjectCount <= 0) {
+			console.error('Attempt to draw 0 objects');
```

Comment:
You could use `onBugIndicatingError`/`onUnexpectedError` here to report a non-throwing error, which makes it through telemetry.
VS Code is so full of expected `console.error`s, that this is like logging to /dev/null.

Reply:
Yes I should definitely remove the `console.error`. I'll have to think about whether we need to do anything at all here, I thought maybe it's nice to know that something is off, but I don't think I've ever hit this during development.

Reply:
Yes, but then leverage the power of telemetry :)